src/main/java/com/nvidia/spark/rapids/jni/kudo/KudoTableMerger.java

```diff
@@ -155,80 +141,64 @@ private static int copyValidityBuffer(HostMemoryBuffer dest, int startBit,
-          curSrcBitIdx = 0;
-          curSrcByteIdx += 1;
-        }
-      } else {
-        // Extract appendCount bits from srcByte, starting from curSrcBitIdx
-        byte mask = (byte) (((1 << leftBitsInCurSrcByte) - 1) & 0xFF);
-        srcByte = (byte) ((srcByte >>> curSrcBitIdx) & mask);
-
-        byte nextSrcByte = src.getByte(curSrcByteIdx + 1);
-        byte nextSrcByteMask = (byte) ((1 << (appendCount - leftBitsInCurSrcByte)) - 1);
-        nextSrcByte = (byte) (nextSrcByte & nextSrcByteMask);
-        nextSrcByte = (byte) (nextSrcByte << leftBitsInCurSrcByte);
-        srcByte = (byte) (srcByte | nextSrcByte);
+      int curSrcOffset = srcOffset + (curSrcIdx / 32) * Integer.BYTES;
+      int curSrcBitIdx = curSrcIdx % 32;
 
-        nullCount += (appendCount - NUMBER_OF_ONES[srcByte & 0xFF]);
+      // This is safe since we always have validity buffer 4 bytes padded
+      int srcInt = src.getInt(curSrcOffset);
+      srcInt = srcInt >>> curSrcBitIdx;
 
-        // Sets the bits in destination buffer starting from curDestBitIdx to 0
-        byte destByte = dest.getByte(curDestByteIdx);
-        destByte = (byte) (destByte & ((1 << curDestBitIdx) - 1));
+      if (dest.getLength() >= (curDestOffset + Integer.BYTES)) {
+        // We have enough room to get an int
+        int destInt = dest.getInt(curDestOffset);
+        destInt &= (1 << curDestBitIdx) - 1;
+        destInt |= srcInt << curDestBitIdx;
+        dest.setInt(curDestOffset, destInt);
 
-        // Update destination byte with the bits from source byte
-        destByte = (byte) (destByte | (srcByte << curDestBitIdx));
-        dest.setByte(curDestByteIdx, destByte);
-
-        // Update the source byte index and bit index
-        curSrcByteIdx += 1;
-        curSrcBitIdx = appendCount - leftBitsInCurSrcByte;
-      }
+        int appendCount = min(leftRowCount, 32 - Math.max(curSrcBitIdx, curDestBitIdx));
 
-      curIdx += appendCount;
-
-      // Update the destination byte index and bit index
-      curDestBitIdx += appendCount;
-      if (curDestBitIdx == 8) {
-        curDestBitIdx = 0;
-        curDestByteIdx += 1;
+        curDestIdx += appendCount;
+        curSrcIdx += appendCount;
+        if (appendCount == 32) {
+          nullCount += 32 - Integer.bitCount(srcInt);
+        } else {
+          int mask = (1 << appendCount) - 1;
+          nullCount += (appendCount  - Integer.bitCount(srcInt & mask));
+        }
+      } else {
+        int destBufRemBytes = toIntExact(dest.getLength() - curDestOffset);
+        byte[] destBytes = new byte[4];
+        dest.getBytes(destBytes, 0, curDestOffset, destBufRemBytes);
+        int destInt = ByteBuffer.wrap(destBytes).order(ByteOrder.LITTLE_ENDIAN).getInt();
+        destInt &= (1 << curDestBitIdx) - 1;
+        destInt |= srcInt << curDestBitIdx;
+
+        ByteBuffer.wrap(destBytes).order(ByteOrder.LITTLE_ENDIAN).putInt(destInt);
+        dest.setBytes(curDestOffset, destBytes, 0, destBufRemBytes);
```

Comment:
Similar to above, can call setInt here and leverage ByteOrder to determine when we need to swap bytes or not.